rfcs/0008-distributed-dns.md

```diff
@@ -0,0 +1,270 @@
+
+## Switching strategy (IE from LoadBalanced to Simple or the other way)
+The DNSPolicy offers 2 strategies 
+1) Simple (one A record multiple values)
+2) LoadBalanced (leverages CNAMES and A records to provide advanced DNS based routing to be used (example GEO and Weighted) )
+
+The strategy field will be immutable once set. To migrate from one strategy to the other you must first delete the existing policy and recreate it with a new strategy. In a multi-cluster environment, this can mean you have a period of time where 2 different strategies are set (simple on one and load balanced on another). The controller will mitigate against this in the following way. When attempting to update the remote zone, it first does a read. If it spots existing records in that zone that correlate with a different strategy, it will not write its own records but will flag this as an error state in the status. So the records will remain in the original strategy until all policies for that listener host match (more in this below).
+
+
+## Orphan Mitigation
+
+It is possible for an orphan set of records to be created if for example a cluster is removed and the controller on that cluster is not given time to clean up. This is not something we currently mitigate against directly in the controller. However, you can define health checks that would remove these orphans from the DNS response. Additionally, there is future work that leverages a "heart beat" that should allow each alive controller to see and remove a dead controllers records within a given criteria.
+
+### Health Checks
+The health checks provided by the various DNS Providers will be used AWS will be the only one implemented initially as part of this RFC as we have done it before. For the others we may choose to cover this area in more detail in a subsequent RFC and they may be limited to having infra within that provider.
+
+[AWS](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/DNS-failover.html) (well understood)
+[Google DNS](https://cloud.google.com/DNS/docs/zones/manage-routing-policies#before_you_begin)
+[Azure](https://learn.microsoft.com/en-us/azure/traffic-manager/traffic-manager-monitoring)
+
+## OCM / ACM
+This is very similar, with the caveat that the "cluster" is the hub cluster for this OCM and that the gateway will be a Kuadrant gatewayClass. You can think about this setup as "single cluster" with a gateway that has multiple addresses. It will continue to work with these changes. 
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+
+## General Logic Overview
+
+### Kuadrant Operator
+
+The general flow in the Kuadrant operator follows a single path, where it will always output a valid DNSRecord, which specifies everything needed for the listener host of a given gateway on the cluster, and is unaware of the concept of distributed DNS. The core logic for doing this already exists.
```

Comment:
what if DNSRecord object created from DNSPolicy is manipulated directly? Does kuadrant-operator reconcile it back to the desired state? if not and "multicluster is active" I think the dns-operator can easily be confused as the record structure may not be compliant with either strategy.

Reply:
Yes this will be part of reconciliation. IE watch the DNSRecord and bring back to the right state